nncf/quantization/algorithms/weight_compression/algorithm.py

```diff
@@ -541,125 +547,134 @@ def _get_activation_node_and_port(self, node: NNCFNode, nncf_graph: NNCFGraph) -
-    ) -> Dict[str, List[Tensor]]:
+        if statistic_points is not None:
+            self._statistics = self._get_statistics(matmul_input_to_output_nodes_map, statistic_points)
+
+    def get_statistic_points(
+        self,
+        model: TModel,
+        graph: NNCFGraph,
+        nodes_and_port_ids: Iterable[Tuple[NNCFNode, int]],
+        subset_size: Optional[int] = None,
+    ) -> StatisticPointsContainer:
         """
-        Collects input activations for the given nodes on the dataset.
+        Returns statistic points, for which StatisticsCollector should collect statistics.
 
-        :param dataset: Dataset to collect values.
-        :param subset_size: Number of data samples to calculate activation statistics used for assigning different
-            quantization precision.
-        :param nodes_to_compress: List of nodes, whose inputs are collected.
         :param model: Model for statistics collection.
         :param graph: Model graph.
-        :return: statistics values itself per node name.
+        :param nodes_and_port_ids: Nodes and port ids for which statistics should be collected.
+        :param subset_size: Number of samples to collect.
+        :return: Statistic points, for which StatisticsCollector should collect statistics.
         """
-        activations = {}
-        _collected_stat_inputs_map = {}
-        statistic_container = StatisticPointsContainer()
-        all_act_nodes = set()
-        act_vs_shared_node_names_mapping = defaultdict(list)
-        matmul_metatypes = self._backend_entity.matmul_metatypes
-        filtered_nodes = filter(lambda node: node.metatype in matmul_metatypes, nodes_to_compress)
-        for node in filtered_nodes:
-            act_node, output_port_id = self._get_activation_node_and_port(node, graph)
-            act_node_name = act_node.node_name
-            if act_node_name in all_act_nodes:
-                act_vs_shared_node_names_mapping[act_node_name].append(node.node_name)
-                continue
-            all_act_nodes.add(act_node_name)
-            output_id = (act_node_name, output_port_id)
-            _collected_stat_inputs_map[node.node_name] = output_id
 
+        statistic_container = StatisticPointsContainer()
+        for node, output_port_id in nodes_and_port_ids:
             statistic_point = self._backend_entity.target_point(
-                TargetType.POST_LAYER_OPERATION, act_node_name, port_id=output_port_id
+                TargetType.POST_LAYER_OPERATION, node.node_name, port_id=output_port_id
+            )
+            # Each activation tensor is assumed to have shape [B, L, C], where B=1 and L is a sequence length dimension.
+            # We reduce activations across B and L dimensions, mean is used as a reduction function.
+            stat_collector = self._backend_entity.mean_statistic_collector(
+                reduction_axes=(0, 1), subset_size=subset_size
```

Comment:
Stable Diffusion 1.5 has a time embedding projection with 2D shapes.
What do you think about this way of generalization: "reduction axes are all except the last one, which is hidden dimension"?
![image](https://github.com/user-attachments/assets/f75e4c88-4f66-4d17-b4b0-f989e41c5db8)

Reply:
Also there're 4D activations for a few matmuls in chatglm2 and dolly2-3b, but similar to text-embeddings in stable diffusion it's open question how to properly compress them.   

![image](https://github.com/user-attachments/assets/8534e559-cb69-42f8-a380-9b2d6795de44)
![image](https://github.com/user-attachments/assets/1c5e2082-72d0-44db-91fa-a72e800726b2)

https://pytorch.org/docs/stable/generated/torch.nn.functional.scaled_dot_product_attention.html
![image](https://github.com/user-attachments/assets/63bbdb19-7b45-4a97-beca-86fe529f9a33)
![image](https://github.com/user-attachments/assets/ea95c229-7039-4c33-b50f-5f0680c35a33)

Reply:
Decided to address this in a separate PR